File: 75 days Challenge Leetcode/Stack/Day-25/Solution.java
import java.util.*;

class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();  // Stack to keep track of surviving asteroids

        for (int asteroid : asteroids) {
            boolean destroyed = false;  // Flag to track if the current asteroid is destroyed

            // Check for collisions between the current asteroid and those in the stack
            while (!stack.isEmpty() && asteroid < 0 && stack.peek() > 0) {
                if (stack.peek() < Math.abs(asteroid)) {
                    stack.pop(); // The asteroid in the stack is destroyed
                    continue;
                } else if (stack.peek() == Math.abs(asteroid)) {
                    stack.pop(); // Both asteroids are destroyed
                }
                
                destroyed = true; // Mark the current asteroid as destroyed
                break;
            }
            if (!destroyed) stack.push(asteroid); // If not destroyed, push the asteroid onto the stack
        }

        // Prepare the result array to collect remaining asteroids
        int[] result = new int[stack.size()];
        for (int i = stack.size() - 1; i >= 0; i--) { // Reverse to maintain original order
            result[i] = stack.pop();
        }

        return result;  // Return the result array
    }
}

File: 75 days Challenge Leetcode/Stack/Day-25/Solution.cpp
#include <stack>  
#include <vector>  

class Solution {  
public:  
    vector<int> asteroidCollision(vector<int>& asteroids) {  
        stack<int> st;  // Stack to keep track of surviving asteroids

        for (int asteroid : asteroids) {  
            bool destroyed = false;  // Flag to track if the current asteroid is destroyed

            // Check for collisions between the current asteroid and those in the stack
            while (!st.empty() && asteroid < 0 && st.top() > 0) {  
                if (st.top() < abs(asteroid)) {  
                    st.pop(); // The asteroid in the stack is destroyed
                    continue;  
                } else if (st.top() == abs(asteroid)) {
                    st.pop(); // Both asteroids are destroyed
                }
                
                destroyed = true; // Mark the current asteroid as destroyed
                break;  
            }  
            if (!destroyed) st.push(asteroid);  // If not destroyed, push the asteroid onto the stack
        }  

        vector<int> result;  
        while (!st.empty()) {  
            result.push_back(st.top());  // Collect remaining asteroids in result
            st.pop();  
        }  

        reverse(result.begin(), result.end()); // Reverse the result to maintain original order
        return result;  
    }  
};  
